File: src/services/audit_service.py
import sqlite3
from datetime import datetime, timezone
import getpass


def log(conn: sqlite3.Connection, action: str, table_name: str, record_id: str,
         old_value: str | None = None, new_value: str | None = None, user: str | None = None) -> None:
    conn.execute(
        """INSERT INTO audit_log (timestamp, user, action, table_name, record_id, old_value, new_value)
           VALUES (?, ?, ?, ?, ?, ?, ?)""",
        (
            datetime.now(timezone.utc).isoformat(),
            user or getpass.getuser(),
            action,
            table_name,
            record_id,
            old_value,
            new_value,
        ),
    )

File: docs/src/services/audit-service.js
// Port of src/services/audit_service.py
import { persistence } from "../db/persistence.js";

export function log(action, tableName, recordId, { oldValue = null, newValue = null, user = null } = {}) {
  persistence.run(
    `INSERT INTO audit_log (timestamp, user, action, table_name, record_id, old_value, new_value)
     VALUES (?, ?, ?, ?, ?, ?, ?)`,
    [new Date().toISOString(), user || "browser-user", action, tableName, recordId, oldValue, newValue]
  );
}
